Validate datastore engine

Himanshu372 opened the pull request:
## Description
<!-- Provide a detailed description of the changes -->
### Added 
1. Added a `DatastoreEngine` type which takes up const permissible engine type values
2. Added helper methods for the type `DatastoreEngine`
3. Added checks to validate engine AFTER it is marshalled into config type
4. Added tests 

## References
Addresses comment - https://github.com/openfga/openfga/pull/712#issuecomment-1574015631
Closes https://github.com/openfga/openfga/issues/681

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

Himanshu372 commented:
Hi @jon-whit, based on our last discussion, raised this PR to verify Datastore engine. Please have a look. Thanks

jon-whit commented on cmd/shared.go:
Mind renaming to `cmd/cmd.go`?

jon-whit commented on cmd/shared.go:
```diff
@@ -0,0 +1,40 @@
+// IsValid checks if DatastoreEngine e is valid
+func (e DatastoreEngine) IsValid() bool {
+	var isValid bool
+	for _, engineType := range []DatastoreEngine{Memory, Postgres, MySQL} {
```
This code can change to simply:

```go
for _, engine := range []DatastoreEngine{Memory, Postgres, MySQL} {
    if engine.String() == e.String() {
        return true
    }
}

return false
```

The variable assignment and then return is redundant.

jon-whit commented on cmd/shared.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+// NewDatastoreEngine inits a new valid DatastoreEngine type
+func NewDatastoreEngine(engine string) (*DatastoreEngine, error) {
```
Why return a pointer instead of a value type here?

jon-whit commented on cmd/run/run_test.go:
```diff
@@ -283,6 +284,23 @@ func TestVerifyConfig(t *testing.T) {
+
+	t.Run("correct datastore engine specified", func(t *testing.T) {
+		cfg := DefaultConfig()
+		cfg.Datastore.URI = "postgresql://localhost:5432"
```
I didn't think `postgresql` was a valid prefix? The enumerations in `cmd/shared.go` below define `postgres`.

This test should return an error, and we should change this test to assert `postgres://localhost:5432` does not.

Himanshu372 replied:
I think, I accidently set URI in this test. I just wanted to test for correct DatastoreEngine. After changes, I am getting expected result. 
Although one thing which I noticed is, on setting wrong postgres URI, `RunServer` stalls and waits for connection. I think we should fail early. Let me know your thoughts

jon-whit reviewed (changes requested):
Left some comments for review.

Himanshu372 commented:
> Left some comments for review.

@jon-whit Addressed comments, plz check.

Himanshu372 commented:
> > Left some comments for review.
> 
> @jon-whit Addressed comments, plz check.

Hey @jon-whit plz have a look and let me know if the changes are OK

miparnisari commented on cmd/cmd.go:
```diff
@@ -0,0 +1,39 @@
+func NewDatastoreEngine(engine string) (DatastoreEngine, error) {
+	dsEngine := DatastoreEngine(engine)
+	if isValid := dsEngine.IsValid(); !isValid {
+		return "", fmt.Errorf("invalid datastore engine '(%s)'", engine)
```
We have similar code here https://github.com/openfga/openfga/blob/0f3e337dec1503fa3fc6b045f2efbb27a4defa19/cmd/run/run.go#L360 and here https://github.com/openfga/openfga/blob/8dc2ed62413e7cccc7da57da057f1f1fc17dcb8a/cmd/validatemodels/validate_models.go#L70

miparnisari commented:
Hi @Himanshu372, i apologize for taking so long to review this. I took the liberty of updating this PR with respect to main. I also added some more validations to the datastore configs.

curfew-marathon commented:
Thank you for opening the PR.  At this time we are closing the PR as it has gone stale and needs to be brought up to date.  In the future feel free to re-open the PR.  Thanks again.
